File: 1/main.py
import os
import requests
from bs4 import BeautifulSoup
from urllib.parse import urljoin
from loguru import logger


class PageDownloader:
    @staticmethod
    def download(url):
        try:
            response = requests.get(url)
            if response.status_code == 200:
                return response.text
        except requests.exceptions.RequestException:
            logger.info(f"Failed to fetch {url}")
            return None
        return None


class TextExtractor:
    @staticmethod
    def extract(html):
        soup = BeautifulSoup(html, 'html.parser')
        text = ' '.join(soup.stripped_strings)
        return text


class TextSaver:
    @staticmethod
    def save(text, filename):
        with open(filename, 'w', encoding='utf-8') as file:
            file.write(text)


class WebCrawler:
    def __init__(self, *,  urls, target_dir, max_pages=100, min_words=1000):
        self.urls = urls
        self.target_dir = target_dir
        self.max_pages = max_pages
        self.min_words = min_words
        self.__crawled_pages = set()

    def crawl(self):
        index_file = open('index.txt', 'w', encoding='utf-8')
        pages_to_crawl = self.urls.copy()

        while len(self.__crawled_pages) < self.max_pages and pages_to_crawl:
            url = pages_to_crawl.pop(0)
            html = PageDownloader.download(url)
            if any(
                    [
                        url in self.__crawled_pages,
                        not html
                     ]
            ):
                continue

            text = TextExtractor.extract(html)
            if not self.is_sufficient_text(text):
                continue

            self.save_page(url, text, index_file)
            self.add_links_to_crawl(url, html, pages_to_crawl)

        index_file.close()

    def is_sufficient_text(self, text):
        return len(text.split()) >= self.min_words

    def save_page(self, url, text, index_file):
        filename = f"page_{len(self.__crawled_pages) + 1}.txt"
        TextSaver.save(text, os.path.join(self.target_dir, filename))
        index_file.write(f"{len(self.__crawled_pages) + 1}: {url}\n")
        self.__crawled_pages.add(url)

    def add_links_to_crawl(self, url, html, pages_to_crawl):
        soup = BeautifulSoup(html, 'html.parser')
        links = [link.get('href') for link in soup.find_all('a') if link.get('href')]
        for link in links:
            absolute_link = urljoin(url, link)
            if absolute_link not in self.__crawled_pages:
                pages_to_crawl.append(absolute_link)


if __name__ == "__main__":
    target_directory = "./downloaded_pages"
    os.makedirs(target_directory, exist_ok=True)
    crawler = WebCrawler(
        urls=["https://dev-ed.ru/blog/101-python3-qa/#32", "https://habr.com/ru/feed"],
        target_dir=target_directory
    )
    crawler.crawl()

File: 1/main.go
package main

import (
	"fmt"
	"github.com/PuerkitoBio/goquery"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
)

type WebCrawler struct {
	URLs       []string
	TargetDir  string
	MaxPages   int
	MinWords   int
	CrawledSet map[string]bool
}

func NewWebCrawler(urls []string, targetDir string, maxPages, minWords int) *WebCrawler {
	return &WebCrawler{
		URLs:       urls,
		TargetDir:  targetDir,
		MaxPages:   maxPages,
		MinWords:   minWords,
		CrawledSet: make(map[string]bool),
	}
}

func (wc *WebCrawler) Crawl() {
	indexFile, err := os.Create(filepath.Join(wc.TargetDir, "index.txt"))
	if err != nil {
		log.Fatalf("Failed to create index file: %v", err)
	}
	defer indexFile.Close()

	for len(wc.CrawledSet) < wc.MaxPages && len(wc.URLs) > 0 {
		url := wc.URLs[0]
		wc.URLs = wc.URLs[1:]

		if wc.CrawledSet[url] {
			continue
		}

		html, err := wc.download(url)
		if err != nil {
			log.Printf("Failed to fetch %s: %v", url, err)
			continue
		}

		text := wc.extract(html)
		if !wc.isSufficientText(text) {
			continue
		}

		filename := fmt.Sprintf("page_%d.txt", len(wc.CrawledSet)+1)
		filePath := filepath.Join(wc.TargetDir, filename)
		err = os.WriteFile(filePath, []byte(text), 0644)
		if err != nil {
			log.Printf("Failed to save page %s: %v", url, err)
			continue
		}
		indexFile.WriteString(fmt.Sprintf("%d: %s\n", len(wc.CrawledSet)+1, url))
		wc.CrawledSet[url] = true

		wc.addLinksToCrawl(url, html, &wc.URLs)
	}
}

func (wc *WebCrawler) download(url string) (string, error) {
	resp, err := http.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("failed to fetch %s: %s", url, resp.Status)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}

	return string(body), nil
}

func (wc *WebCrawler) extract(html string) string {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		log.Printf("Failed to parse HTML: %v", err)
		return ""
	}
	return doc.Text()
}

func (wc *WebCrawler) isSufficientText(text string) bool {
	return len(strings.Fields(text)) >= wc.MinWords
}

func (wc *WebCrawler) joinURL(baseURL, link string) (string, error) {
	base, err := url.Parse(baseURL)
	if err != nil {
		return "", err
	}
	rel, err := url.Parse(link)
	if err != nil {
		return "", err
	}
	resolvedURL := base.ResolveReference(rel).String()
	return resolvedURL, nil
}
func (wc *WebCrawler) addLinksToCrawl(url, html string, pagesToCrawl *[]string) {
	doc, err := goquery.NewDocumentFromReader(strings.NewReader(html))
	if err != nil {
		log.Printf("Failed to parse HTML: %v", err)
		return
	}

	doc.Find("a").Each(func(_ int, s *goquery.Selection) {
		link, exists := s.Attr("href")
		if !exists {
			return
		}
		absoluteLink, err := wc.joinURL(url, link)
		if err != nil {
			log.Fatal(err)
		}
		if !wc.CrawledSet[absoluteLink] {
			*pagesToCrawl = append(*pagesToCrawl, absoluteLink)
		}
	})
}

func main() {
	dir, _ := os.Getwd()
	targetDirectory := filepath.Join(dir, "downloaded_pages_uk_rf")
	fmt.Printf(targetDirectory)
	os.MkdirAll(targetDirectory, os.ModePerm)
	crawler := NewWebCrawler(
		[]string{"https://habr.com/ru/feed"},
		targetDirectory,
		100,
		1000,
	)
	crawler.Crawl()
}
